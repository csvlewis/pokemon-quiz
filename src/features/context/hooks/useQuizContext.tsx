import { useContext } from "react";
import { QuizContextType } from "@/features/context/types/QuizContextType";
import { QuizContext } from "@/features/context/components/QuizContext";

export function useQuizContext(): QuizContextType {
  return useContext<QuizContextType>(QuizContext);
}
