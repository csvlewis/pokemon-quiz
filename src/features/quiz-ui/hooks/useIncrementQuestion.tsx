import { useQuizContext } from "@/features/context/components/QuizContextProvider";

export function useIncrementQuestion(): () => void {
  const {
    currentQuestion,
    questions,
    setCurrentQuestion,
    setDisplayAnswer,
    setSelectedAnswer,
    setShowResults,
    setQuizInProgress,
  } = useQuizContext();

  return () => {
    if (currentQuestion >= questions.length) {
      setQuizInProgress(false);
      setShowResults(true);
      return;
    }
    setCurrentQuestion(currentQuestion + 1);
    setDisplayAnswer(false);
    setSelectedAnswer({});
  };
}
