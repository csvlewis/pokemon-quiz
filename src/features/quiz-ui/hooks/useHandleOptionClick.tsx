import { useQuizContext } from "@/features/context/hooks/useQuizContext";
import { Option } from "@/types/Option";

export function useHandleOptionClick(): (option: Option) => void {
  const { displayAnswer, selectedAnswer, setSelectedAnswer } = useQuizContext();

  return (option: Option) => {
    if (displayAnswer) return;
    if (option === selectedAnswer) {
      setSelectedAnswer({});
    } else {
      setSelectedAnswer(option);
    }
  };
}
