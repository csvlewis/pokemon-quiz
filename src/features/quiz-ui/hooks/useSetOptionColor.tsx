import { OverridableStringUnion } from "@mui/types";
import { useQuizContext } from "@/features/context/hooks/useQuizContext";
import { Option } from "@/types/Option";

export function useSetOptionColor(): (
  option: Option
) => OverridableStringUnion<"error" | "primary" | "success"> {
  const { displayAnswer } = useQuizContext();

  return (option: Option) => {
    let color: OverridableStringUnion<"error" | "primary" | "success"> =
      "primary";

    if (displayAnswer) {
      if (!option.isCorrect) {
        color = "error";
      } else {
        color = "success";
      }
    }

    return color;
  };
}
