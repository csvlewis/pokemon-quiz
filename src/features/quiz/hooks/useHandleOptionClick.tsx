import { Option } from "@/types/Option";
import { useQuizContext } from "@/features/context/hooks/useQuizContext";

export function useHandleOptionClick(): (option: Option) => void {
  const { displayAnswer, selectedAnswer, setSelectedAnswer } = useQuizContext();

  return (option: Option) => {
    if (displayAnswer) return;
    if (option === selectedAnswer) {
      setSelectedAnswer({ text: "", isCorrect: false });
    } else {
      setSelectedAnswer(option);
    }
  };
}
