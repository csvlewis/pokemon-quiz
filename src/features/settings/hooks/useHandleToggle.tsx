import { useQuizContext } from "@/features/context/hooks/useQuizContext";

export function useHandleToggle(): (value: string) => void {
  const { selectedQuestionTypes, setSelectedQuestionTypes } = useQuizContext();

  return (value: string) => {
    const currentIndex = selectedQuestionTypes.indexOf(value);
    const newQuestionTypes = [...selectedQuestionTypes];

    if (currentIndex === -1) {
      newQuestionTypes.push(value);
    } else {
      newQuestionTypes.splice(currentIndex, 1);
    }

    setSelectedQuestionTypes(newQuestionTypes);
  };
}
