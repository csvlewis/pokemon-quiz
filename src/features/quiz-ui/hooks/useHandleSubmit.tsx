import { useQuizContext } from "@/features/context/hooks/useQuizContext";
import { useIncrementQuestion } from "@/features/quiz-ui/hooks/useIncrementQuestion";

export function useHandleSubmit(): () => void {
  const {
    correctAnswers,
    displayAnswer,
    selectedAnswer,
    setCorrectAnswers,
    setDisplayAnswer,
  } = useQuizContext();
  const incrementQuestion = useIncrementQuestion();

  return () => {
    if (displayAnswer) {
      incrementQuestion();
    } else {
      if (selectedAnswer.isCorrect) setCorrectAnswers(correctAnswers + 1);
      setDisplayAnswer(true);
    }
  };
}
