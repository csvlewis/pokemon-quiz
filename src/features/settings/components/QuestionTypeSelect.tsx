import {
  Checkbox,
  List,
  ListItem,
  ListItemButton,
  ListItemIcon,
  ListItemText,
  Stack,
  Typography,
} from "@mui/material";

import { questionTypes } from "@/config/questionTypes";
import { useQuizContext } from "@/features/context/hooks/useQuizContext";
import { useHandleToggle } from "@/features/settings/hooks/useHandleToggle";

export function QuestionTypeSelect(): React.ReactNode {
  const { selectedQuestionTypes } = useQuizContext();
  const handleToggle: (value: string) => void = useHandleToggle();

  return (
    <Stack alignItems="center">
      <Typography>Select at least one question type for your quiz:</Typography>
      <List sx={{ width: "100%" }}>
        {questionTypes.map((questionType, index) => {
          return (
            <ListItem key={index} disablePadding>
              <ListItemButton
                onClick={() => handleToggle(questionType.name)}
                dense
                disableRipple
              >
                <ListItemIcon>
                  <Checkbox
                    checked={selectedQuestionTypes.includes(questionType.name)}
                    disableRipple
                  />
                </ListItemIcon>
                <ListItemText
                  primary={`${questionType.name} - ${questionType.description}`}
                />
              </ListItemButton>
            </ListItem>
          );
        })}
      </List>
    </Stack>
  );
}
