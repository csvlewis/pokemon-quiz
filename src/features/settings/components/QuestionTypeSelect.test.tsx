import React from "react";
import { render } from "@testing-library/react";
import { useQuizContext } from "@/features/context/hooks/useQuizContext";
import { QuestionTypeSelect } from "./QuestionTypeSelect";

jest.mock("@/features/context/hooks/useQuizContext");

const mockedUseQuizContext = useQuizContext as jest.Mock;

describe("QuestionTypeSelect", () => {
  it("should render correctly", () => {
    mockedUseQuizContext.mockReturnValue({
      selectedQuestionTypes: [],
    });

    const { container } = render(<QuestionTypeSelect />);
    expect(container).toMatchSnapshot();
  });
});
