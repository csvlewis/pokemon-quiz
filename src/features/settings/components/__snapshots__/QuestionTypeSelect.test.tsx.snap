// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionTypeSelect should render correctly 1`] = `
<div>
  <div
    class="MuiStack-root css-163ljxj-MuiStack-root"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
    >
      Select at least one question type for your quiz:
    </p>
    <ul
      class="MuiList-root MuiList-padding css-v54zg0-MuiList-root"
    >
      <li
        class="MuiListItem-root MuiListItem-gutters css-19f6boz-MuiListItem-root"
      >
        <div
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters css-2knotq-MuiButtonBase-root-MuiListItemButton-root"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemIcon-root css-cokf1l-MuiListItemIcon-root"
          >
            <span
              class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-rji8ye-MuiButtonBase-root-MuiCheckbox-root"
            >
              <input
                class="PrivateSwitchBase-input css-j8yymo"
                data-indeterminate="false"
                type="checkbox"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="CheckBoxOutlineBlankIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
            </span>
          </div>
          <div
            class="MuiListItemText-root MuiListItemText-dense css-cfq8qh-MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-7ctihp-MuiTypography-root"
            >
              Name That Pokémon - Identify the Pokémon from its image
            </span>
          </div>
        </div>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters css-19f6boz-MuiListItem-root"
      >
        <div
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters css-2knotq-MuiButtonBase-root-MuiListItemButton-root"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemIcon-root css-cokf1l-MuiListItemIcon-root"
          >
            <span
              class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-rji8ye-MuiButtonBase-root-MuiCheckbox-root"
            >
              <input
                class="PrivateSwitchBase-input css-j8yymo"
                data-indeterminate="false"
                type="checkbox"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="CheckBoxOutlineBlankIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
            </span>
          </div>
          <div
            class="MuiListItemText-root MuiListItemText-dense css-cfq8qh-MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-7ctihp-MuiTypography-root"
            >
              Pokémon Ability - Choose the correct ability for the Pokémon
            </span>
          </div>
        </div>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters css-19f6boz-MuiListItem-root"
      >
        <div
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters css-2knotq-MuiButtonBase-root-MuiListItemButton-root"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemIcon-root css-cokf1l-MuiListItemIcon-root"
          >
            <span
              class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-rji8ye-MuiButtonBase-root-MuiCheckbox-root"
            >
              <input
                class="PrivateSwitchBase-input css-j8yymo"
                data-indeterminate="false"
                type="checkbox"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="CheckBoxOutlineBlankIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
            </span>
          </div>
          <div
            class="MuiListItemText-root MuiListItemText-dense css-cfq8qh-MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-7ctihp-MuiTypography-root"
            >
              Pokémon Move - Identify the move that the Pokémon can learn
            </span>
          </div>
        </div>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters css-19f6boz-MuiListItem-root"
      >
        <div
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters css-2knotq-MuiButtonBase-root-MuiListItemButton-root"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemIcon-root css-cokf1l-MuiListItemIcon-root"
          >
            <span
              class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-rji8ye-MuiButtonBase-root-MuiCheckbox-root"
            >
              <input
                class="PrivateSwitchBase-input css-j8yymo"
                data-indeterminate="false"
                type="checkbox"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="CheckBoxOutlineBlankIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
            </span>
          </div>
          <div
            class="MuiListItemText-root MuiListItemText-dense css-cfq8qh-MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-7ctihp-MuiTypography-root"
            >
              Pokémon Type - Choose the correct type for the Pokémon
            </span>
          </div>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
